tls/extensions/s2n_client_psk.c

```diff
@@ -210,53 +208,20 @@ static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn,
     ENSURE_REF(conn);
     ENSURE_REF(wire_identities_in);
 
-    uint16_t identities_count = 0;
-    GUARD_RESULT(s2n_count_psk_identities(wire_identities_in, &identities_count));
-    ENSURE_GT(identities_count, 0);
-    ENSURE_LTE(identities_count, MAX_NUM_OF_PSK_IDENTITIES);
+    struct s2n_offered_psk_list identity_list = { .wire_data = *wire_identities_in };
 
-    DEFER_CLEANUP(struct s2n_blob wire_identities_blob = { 0 }, s2n_free);
-    GUARD_AS_RESULT(s2n_alloc(&wire_identities_blob, identities_count * sizeof(struct s2n_psk_identity)));
-    struct s2n_psk_identity *wire_identities = (struct s2n_psk_identity*)(void*) wire_identities_blob.data;
-
-    uint16_t wire_index = 0;
-    while (s2n_stuffer_data_available(wire_identities_in) > 0) {
-        uint16_t identity_size = 0;
-        GUARD_AS_RESULT(s2n_stuffer_read_uint16(wire_identities_in, &identity_size));
-        ENSURE_GT(identity_size, 0);
-
-        uint8_t *identity_data = s2n_stuffer_raw_read(wire_identities_in, identity_size);
-        ENSURE_REF(identity_data);
-
-        wire_identities[wire_index].data = identity_data;
-        wire_identities[wire_index].length = identity_size;
-
-        /**
-         *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
-         *# For identities established externally, an obfuscated_ticket_age of 0 SHOULD be
-         *# used, and servers MUST ignore the value.
-         */
-        uint32_t obfuscated_ticket_age = 0;
-        GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
+    if (conn->config->psk_selection_cb) {
```

Comment:
This flow worries me.

Let's say that a server supports PSKs (and thus implements `psk_selection_cb`) and also support session resumption.  Consider the following flows:
1. Client attempts initial connection and includes an external PSK
2. Server uses `psk_selection_cb` to select the external PSK. (Connection succeeds)
3. Server returns a session resumption PSK.
4. Client attempts a new connection and includes both the session PSK and the external PSK
5.  Server uses `psk_selection_cb` to select the external PSK. (Connection succeeds with external PSK.)
    NOTE: It is impossible to select the session PSK because `psk_selection_cb` (correctly) doesn't know about the sesion PSKs. This could be an issue if the session PSK is in position 0 and 0 RTT data is provided.

Consider if (for some strange reason) the client attempts a connection only with the session PSK. The session PSK may still be valid, but since `psk_selection_cb` is non-null, the first branch is taken and we never call `s2n_select_psk_identity`. Would this result in never checking/accepting the session PSK?

Reply:
We talked about this offline.

Basically, psk_selection_cb will support session resumption PSKs once S2N supports session resumption PSKs. All psk_selection_cb needs to select a PSK is the wire index, which will also allow it to select session resumption PSKs. The secrets for external PSKs are configured separately (but that configuration can be triggered from the callback) and the resumption PSK configuration will also happen separately. The psk_selection_cb callback doesn't need to know the secret associated with a PSK to choose that PSK.